File: kotlin-runtime/src/main/kotlin/rosettarpc/RpcMethodDescriptor.kt
package rosettarpc

import com.google.protobuf.Message
import com.google.protobuf.Parser

/**
 * Identifies one RPC method: a service id, a method id, and the parsers needed to decode a
 * request/response of that method's types.
 */
class RpcMethodDescriptor<Req : Message, Resp : Message>(
    val serviceId: String,
    val methodId: String,
    val requestParser: Parser<Req>,
    val responseParser: Parser<Resp>,
)

File: csharp-runtime/RpcMethodDescriptor.cs
using Google.Protobuf;

namespace RosettaRpc;

/// <summary>
/// Identifies one RPC method and carries the typed parsers needed to
/// encode/decode it. Const-constructible and never references any
/// project-specific generated type.
/// </summary>
public sealed class RpcMethodDescriptor<TReq, TResp>
    where TReq : IMessage<TReq>
    where TResp : IMessage<TResp>
{
    public string ServiceId { get; }
    public string MethodId { get; }
    public MessageParser<TReq> RequestParser { get; }
    public MessageParser<TResp> ResponseParser { get; }

    public RpcMethodDescriptor(
        string serviceId,
        string methodId,
        MessageParser<TReq> requestParser,
        MessageParser<TResp> responseParser)
    {
        ServiceId = serviceId;
        MethodId = methodId;
        RequestParser = requestParser;
        ResponseParser = responseParser;
    }
}
